xapian-letor/tests/api_letor.cc

```diff
@@ -94,26 +167,169 @@ DEFINE_TESTCASE(createfeaturevector, generated)
     TEST(fv.empty());
     mset = enquire.get_mset(0, 10);
     TEST(!mset.empty());
+    TEST_EQUAL(mset.size(), 2);
     fv = fl.create_feature_vectors(mset, Xapian::Query("lions"), db);
     TEST_EQUAL(fv.size(), 2);
     TEST_EQUAL(fv[0].get_fcount(), 19);
     TEST_EQUAL(fv[1].get_fcount(), 19);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectoronevector, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker2",
+				       db_index_one_document);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("tigers"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("tigers"), db);
+    TEST(fv.empty());
+    mset = enquire.get_mset(0, 10);
+    TEST(!mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("tigers"), db);
+    TEST_EQUAL(fv.size(), 1);
+    TEST_EQUAL(fv[0].get_fcount(), 19);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectoronevector_wrongquery, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker3",
+				       db_index_one_document);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("llamas"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("llamas"), db);
+    TEST(fv.empty());
+
+    mset = enquire.get_mset(0, 10);
+    TEST(mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("llamas"), db);
+    TEST_EQUAL(fv.size(), 0);
+    return true;
+}
+
+DEFINE_TESTCASE(createfeaturevectorthree, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker4",
+				       db_index_three_documents);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("score"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("score"), db);
+    TEST(fv.empty());
+
+    mset = enquire.get_mset(0, 10);
+    TEST(!mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("score"), db);
+    TEST_EQUAL(fv.size(), 2);
+    TEST_EQUAL(fv[0].get_fcount(), 19);
+    TEST_EQUAL(fv[1].get_fcount(), 19);
+    return true;
+}
+
+
+DEFINE_TESTCASE(preparetrainingfileonedb, generated)
+{
+    string db_path = get_database_path("apitest_listnet_ranker1",
+				       db_index_one_document);
+    string data_directory = test_driver::get_srcdir() + "/testdata/";
+    string query = data_directory + "queryone.txt";
+    string qrel = data_directory + "qrelone.txt";
+    string training_data = data_directory + "training_data_one_doc.txt";
+    Xapian::prepare_training_file(db_path, query, qrel, 10,
+				  "training_output_data_one_doc.txt");
```

Comment:
That sounds like a lack of error checking somewhere.  Probably failing to check that `ifstream` actually opened a file - you need to call `is_open()` to see if opening the file (via the constructor or an explicit call to the `open()` method) actually worked.

That should be fixed, and a regression testcase added (one which fails before the fix and passes after).

Reply:
Would keep it in a different commit, along with all the checks in cases when one of the qrels are missing. 

so a boolean TEST(if.is_open()) would be needed right after every open right?

Reply:
I was assuming the failure was in `Xapian::prepare_training_file()`, but I think I misunderstood where the problem is.

If the problem was a failure to check that we successfully open an `ifstream` inside the library, then `TEST(...)` wouldn't work for checking for the failure as that's a testsuite specific thing - the library code should throw an exception to signal such a problem.  Then a regression testcase would use `TEST_EXCEPTION()` to check that passing a non-existant training file to `Xapian::prepare_training_file()` causes the appropriate exception to be thrown.  And in that case it would make sense to add that regression testcase in a separate commit - probably even better to make it a separate PR so that we can try to wrap this one up and get it merged.

If it's just that we're not checking that `if1` and `if2` below are actually opened successfully, then `TEST(if1.is_open())` etc below does the trick (and that can go in this PR - as a separate commit makes sense there too).